math/tangent.go

```diff
@@ -1,132 +1,114 @@
-		toRemove = toRemove.Mul(toRemove, m)
+		uv := arith.Abs(v)
 
-		//so toRemove = mPi so
-		// |theta-toRemove| < Pi/2
-		//add 1 to the precision for the up comming squaring
-		x = decimal.WithPrecision(precision+1).Sub(theta, toRemove)
+		// Adjust so we have ceil(v/10) + ctx.Precision, but check for overflows.
+		// 1+((v-1)/10) will be widly incorrect for v == 0, but x/y = 0 iff
+		// x = 0 and y != 0. In this case, -2pi <= x >= 2pi, so we're fine.
+		prec, ok := checked.Add(1+((uv-1)/10), uint64(ctx.Precision))
+		if !ok || prec > maxInt {
+			return nil, false
+		}
+		pctx := decimal.Context{Precision: int(prec)}
+
+		pctx.Mul(x0, Pi(&tmp, pctx), x0)
+
+		ctx.Precision++
+		// so toRemove = m*Pi so |x-toRemove| < Pi/2
+		ctx.Sub(x0, x, x0)
 	} else {
-		//add 1 to the precision for the up comming squaring
-		x = decimal.WithPrecision(precision + 1).Copy(theta)
+		x0.Copy(x)
 	}
-	return x, true
+	return x0, true
 }
 
-//Tan returns the tangent of theta(radians).
-// Input range : -pi/2 <= theta <= pi/2
-// Output range: all real numbers
-// Notes:
-//		Tan(NaN)    -> NaN
-//		Tan(+/-Inf) -> NaN
-func Tan(z *decimal.Big, theta *decimal.Big) *decimal.Big {
-	//here we use the formula
-	//tan(x) = sign(x)*sqrt(1/cos(x)^2-1)
+// Tan returns the tangent, in radians, of x.
+//
+// Range:
+//     Input: -pi/2 <= x <= pi/2
+//     Output: all real numbers
+//
+// Special cases:
+//     Tan(NaN) = NaN
+//     Tan(±Inf) = NaN
+func Tan(z, x *decimal.Big) *decimal.Big {
+	if z.CheckNaNs(x, nil) {
+		return z
+	}
 
-	if theta.IsInf(0) || theta.IsNaN(0) {
+	if x.IsInf(0) {
 		z.Context.Conditions |= decimal.InvalidOperation
 		return z.SetNaN(false)
 	}
 
-	calculatingPrecision := z.Context.Precision + defaultExtraPrecision
-	x, ok := prepareTangentInput(calculatingPrecision, theta)
+	ctx := decimal.Context{Precision: precision(z) + defaultExtraPrecision}
+	x0, ok := prepTan(z, x, ctx)
 	if !ok {
 		z.Context.Conditions |= decimal.InvalidOperation
 		return z.SetNaN(false)
 	}
 
-	// tangent has an asymptote at pi/2 and we'll need more precision as we get closer
-	// the reason we need it is as we approach pi/2 is due to the squaring portion,
-	// it will cause the small value of Cosine to be come extremely small.
-	// we COULD fix it by simply doubling the precision, however,
-	// when the precision gets larger it will be a significant impact
-	// to performance, instead we'll only add the extra precision when we need it
-	// by using the difference to see how much extra precision we need
-	// we'll speed things up by only using a quick compare to see if we need
-	// to do a deeper inspection
-	tmpCalculation := calculatingPrecision
-	if decimal.New(14, 1).CmpAbs(theta) < 0 {
-		piOver2 := Pi(decimal.WithPrecision(calculatingPrecision))
-		piOver2.Quo(piOver2, two)
-		var dd *decimal.Big
-		if theta.Signbit() {
-			dd = decimal.WithPrecision(calculatingPrecision).Add(theta, piOver2)
+	ctx.Precision++
```

Comment:
This increase of precision isn't needed for accuracy. (note if you remove this make sure to update line 112 too)